core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -85,54 +100,57 @@ public Status acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
         return unavailableStatus;
       }
     }
+
+    // Since we have a new set of addresses, we are again at first pass
+    numTf.getAndSet(0);
```

Comment:
I think remove these two lines. Because it is much simpler to count the numTF after first pass, instead of start from address update. That is easier to reason about. We have all the conditions to tell it is first pass so we do not need extra variable. See the code snippet in the later comments.
When we have address update, we reset the index already, (only case is READY connection and address still in the new list, when the READY connection goes IDLE, it will reset the index, then it comes to the same situation), so when the first time index goes to invalid when we iterating the list, we can tell it is first pass complete. 
In case that when we receive the exact same set of addresses and they are all in TF already, it will miss the first pass but it is fine because the it is name resolution is already refreshed before update.